### PR TITLE
Support for custom projections

### PR DESCRIPTION
**Pull request by LHBruneton-C2C**
Install proj4js.
Add API function "registerProjection"
Download projection definition from epsg.io if needed and EPSG prefixed name.
Support projection definition in spec.

### CODE REVIEW

**Comment by LHBruneton-C2C** [comment]
Spec used for testing purpose:
```
{
  "layers": [
    {
      "type": "WMS",
      "url": "https://wxs.ign.fr/jhyvi0fgmnuxvfv0zjzorvdn/geoportail/r/wms",
      "layer": "GEOGRAPHICALGRIDSYSTEMS.PLANIGNV2",
      "tiled": false
    }
  ],
  "size": [600, 400],
  "center": [4, 47],
  "dpi": 200,
  "scale": 10000000,
  "projection": "EPSG:2154"
}
```
Works also with tiled set to true.

**Comment by LHBruneton-C2C** [comment]
Spec used for testing purpose:
```
{
  "layers": [
    {
      "type": "WMS",
      "url": "https://wxs.ign.fr/jhyvi0fgmnuxvfv0zjzorvdn/geoportail/r/wms",
      "layer": "GEOGRAPHICALGRIDSYSTEMS.PLANIGNV2",
      "tiled": true
    }
  ],
  "size": [600, 400],
  "center": [4, 47],
  "dpi": 200,
  "scale": 10000000,
  "projection": "EPSG:3947",
  "projectionDefinition": {
    "name": "EPSG:3947",
    "bbox": [48.0, -4.77, 46.0, 7.63],
    "proj4": "+proj=lcc +lat_1=46.25 +lat_2=47.75 +lat_0=47 +lon_0=3 +x_0=1700000 +y_0=6200000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"
  }
}
```

**Comment by tkohr** [comment]
Nice, works fine and code LGTM.

Just wondering if it might be more intuitive for the spec to put "projection" and "projectionDefinition" in one property (to not be forced to indicate the projection code twice when using a definition).

It could also be useful to document that using "projection" calls epsg.io in the background and providing a "projectionDefinition" avoids this call.

Didn't understand why the bbox indicated here https://epsg.io/3947 and here https://epsg.io/?format=json&q=3947 are not the same.

Last questionmark: Should the center in the spec still be indicated in WGS84 or in the projection used for printing? I like WGS84 as it is easy to read but I could image that clients that do not use WGS84 at all could need the latter.

File: src/shared/projections.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import { register } from 'ol/proj/proj4';
+import proj4 from 'proj4';
+
+export async function search(query) {
+  const response = await fetch('https://epsg.io/?format=json&q=' + query);
+  const json = await response.json();
+
+  const results = json['results'];
+  if (results && results.length > 0) {
+    for (let i = 0, ii = results.length; i < ii; i++) {
```
Comment (tkohr) [inline src/shared/projections.js:12]: What happens if multiple results are returned?

**Comment by LHBruneton-C2C** [comment]
> Nice, works fine and code LGTM.
> 
> Just wondering if it might be more intuitive for the spec to put "projection" and "projectionDefinition" in one property (to not be forced to indicate the projection code twice when using a definition).
> 
Yes I was wondering about that too. Couldn't really decide what would be the most intuitive.
> It could also be useful to document that using "projection" calls epsg.io in the background and providing a "projectionDefinition" avoids this call.
> 
Good call, I'll add it to the doc.
> Didn't understand why the bbox indicated here https://epsg.io/3947 and here https://epsg.io/?format=json&q=3947 are not the same.
> 
I don't know either, but clearly the first one is wrong here.
> Last questionmark: Should the center in the spec still be indicated in WGS84 or in the projection used for printing? I like WGS84 as it is easy to read but I could image that clients that do not use WGS84 at all could need the latter.
>
Maybe the center could have a projection, and if none provided, defaults to WGS84.

File: src/shared/projections.js (hunk 1 above)
Comment (LHBruneton-C2C) [reply src/shared/projections.js:12]: The first truthy one is returned. I don't know how some results could be falsy...
Actually, this part was taken from [Reprojection with EPSG.io Search](https://openlayers.org/en/latest/examples/reprojection-by-code.html)

**Comment by LHBruneton-C2C** [comment]
After some more thinking, I decided to keep projection definitions separate from projection, as they do not serve the same purpose. Projection is for defining the projection of the final printing, and projection definitions allow for registering new projections in proj4. Those projections can later be used anywhere in the spec to allow for layers to be downloaded in a projection different than the one for the map, and then reprojected by OpenLayers.

**Pull request merged**